src/common/compile.ts

```diff
@@ -0,0 +1,493 @@
+import { Node, Program } from "../IR";
+import { expandVariants } from "./expandVariants";
+import { defaultDetokenizer, Plugin, Language } from "./Language";
+import { AddWarning, programToSpine, Spine } from "./Spine";
+import { getType } from "./getType";
+import { stringify } from "./stringify";
+import parse from "../frontend/parse";
+import { MinPriorityQueue } from "@datastructures-js/priority-queue";
+import polygolfLanguage from "../languages/polygolf";
+
+export type OptimisationLevel = "nogolf" | "simple" | "full";
+export type Objective = "bytes" | "chars";
+export interface CompilationOptions {
+  level: OptimisationLevel;
+  objective: Objective | ((x: string | null) => number);
+  getAllVariants?: boolean;
+  skipTypecheck?: boolean;
+  restrictFrontend?: boolean;
+  asciiOnly?: boolean;
+}
+
+export function getObjectiveFunc(options: CompilationOptions) {
+  if (options.objective === "bytes") return byteLength;
+  if (options.objective === "chars") return charLength;
+  return options.objective;
+}
+
+// This is what code.golf uses for char scoring
+// https://github.com/code-golf/code-golf/blob/13733cfd472011217031fb9e733ae9ac177b234b/js/_util.ts#L7
+export const charLength = (str: string | null) => {
+  if (str === null) return Infinity;
+  let i = 0;
+  let len = 0;
+
+  while (i < str.length) {
+    const value = str.charCodeAt(i++);
+
+    if (value >= 0xd800 && value <= 0xdbff && i < str.length) {
+      // It's a high surrogate, and there is a next character.
+      const extra = str.charCodeAt(i++);
+
+      // Low surrogate.
+      if ((extra & 0xfc00) === 0xdc00) {
+        len++;
+      } else {
+        // It's an unmatched surrogate; only append this code unit, in
+        // case the next code unit is the high surrogate of a
+        // surrogate pair.
+        len++;
+        i--;
+      }
+    } else {
+      len++;
+    }
+  }
+
+  return len;
+};
+
+export const byteLength = (x: string | null) =>
+  x === null ? Infinity : Buffer.byteLength(x, "utf-8");
```

Comment:
This is a large file, and `charLength` and `byteLength` are really just utility functions -- should be split off to another file.

Perhaps also split off `shorterBy` too (anything else?) to theme the helper file around objective functions.